### PR TITLE
Feat/ec recover

### PR DESCRIPTION
**Pull request by bajpai244**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

This PR adds support for `ec_recover` [precompile](https://www.evm.codes/precompiled#0x01?fork=shanghai).

Resolves: #665 

## What is the new behavior?

- `ec_recover` precompile is supported at contract address `0x1`.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675)

### CODE REVIEW

**Review by bajpai244 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675)* status: 1 of 6 files reviewed, 6 unresolved discussions (waiting on @ClementWalter and @Eikix)

___
*[`crates/evm/src/precompiles/ec_recover.cairo` line 16 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2EJdGAe3G6Ygc5Syq:-Nm4aMq8ALP5Q8V-vXrk:bgpkbcb) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/precompiles/ec_recover.cairo#L16)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

Open question but should this be EcRecoverImpl of EcRecoverTrait? Since again we are in the precompiles folder
</blockquote></details>

I think Precompile makes sense, we have used the same naming convention.

Because, it does depend on taking the VM as input, but if you think otherwise, then I am fine changing it as well!
___
*[`crates/evm/src/precompiles/ec_recover.cairo` line 22 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2FNyV6IsckqYnBod3:-Nm4aYAJ-pGZqW454EN9:b-dry9mn) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/precompiles/ec_recover.cairo#L22)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

Should this be inlined?
</blockquote></details>

You are correct, I looked the codebase again, looks like the opcodes aren't inlined, I thought they were.

Removed!
___
*[`crates/evm/src/precompiles/ec_recover.cairo` line 56 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2FtM5CFa9O4KRv3vl:-Nm4cGrR4q6z8nukfauj:ba5e2iz) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/precompiles/ec_recover.cairo#L56)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

In the spec they check some range: https://github.com/ethereum/execution-specs/blob/3fe6514f2d9d234e760d11af883a47c1263eff51/src/ethereum/shanghai/vm/precompiled_contracts/ecrecover.py#L48
</blockquote></details>

The reason why behind this is that the precompile should return empty data if it fails { i.e, it should not revert }

Now, in our case, the function we use from the core library accepts the message hash, and the signature { which has u256 values }.

So, the extraction of the points on the curve is handled by the core library function, and if these points aren't in the normal range, the core library recovery function itself should return with error { Option::None }, in which case we return empty data { lemme know if u think I am wrong here }.


___
*[`crates/evm/src/precompiles/ec_recover.cairo` line 62 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2FvzjDe_EVMo5ewQc:-Nm4dU2nE61MVjRxVu1i:bo85ack) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/precompiles/ec_recover.cairo#L62)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

same here https://github.com/ethereum/execution-specs/blob/3fe6514f2d9d234e760d11af883a47c1263eff51/src/ethereum/shanghai/vm/precompiled_contracts/ecrecover.py#L48
</blockquote></details>

addressed in the comment above
___
*[`crates/evm/src/precompiles/lib.cairo` line 16 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2G8ujBExBLq9Xlh2q:-Nm4dWoP7bOTT2v4DHS4:b-v13v1h) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/precompiles/lib.cairo#L16)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

ahh so we can match on a felt? noice

btw we have to match on 0 first because of the forced order of cairo's match?
otherwise we can put the 0 case in the `_ => panic` default case
</blockquote></details>

Yeah, we have to the match on 0 because of forced order!

Also, as of now u can only do match on `enums` and `felt252`, but the ordering thing makes this tricky!

Imagine someone has to do a match for 1, 51, 1001.

They would have to write all cases from 0 -> 10001 and only then can they. use the `_` default case!
___
*[`crates/evm/src/tests/test_precompiles/test_ec_recover.cairo` line 19 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2GTz31x_pLwVS9ToH:-Nm4dv0mEgoSc0PS3Nw8:bnxu1tw) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/tests/test_precompiles/test_ec_recover.cairo#L19)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

We should strive to have re-executable test cases and ideally easily verifiable -> e.g. at least a comprehensive comment explaining how we constructed this test case signature, and later a JS script reproducing this test case / a link to a way for us to verify the behaviour (e.g. link to a geth test, or an ef-test case)
</blockquote></details>

Can a link to evm.codes be fine? Because I just took these from their playground, if this works, then I will add on the comment a link for the same!

**Review by Eikix — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675)* status: 1 of 6 files reviewed, 3 unresolved discussions (waiting on @bajpai244 and @ClementWalter)

___
*[`crates/evm/src/precompiles/ec_recover.cairo` line 16 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2EJdGAe3G6Ygc5Syq:-Nm5Qhgw6zhUbujs911G:b3bl1jx) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/precompiles/ec_recover.cairo#L16)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

I think Precompile makes sense, we have used the same naming convention.

Because, it does depend on taking the VM as input, but if you think otherwise, then I am fine changing it as well!
</blockquote></details>

let's go with this for now
___
*[`crates/evm/src/precompiles/ec_recover.cairo` line 56 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2FtM5CFa9O4KRv3vl:-Nm5QtQs8DUq-1XHBksK:bb6ywmx) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/precompiles/ec_recover.cairo#L56)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

The reason why behind this is that the precompile should return empty data if it fails { i.e, it should not revert }

Now, in our case, the function we use from the core library accepts the message hash, and the signature { which has u256 values }.

So, the extraction of the points on the curve is handled by the core library function, and if these points aren't in the normal range, the core library recovery function itself should return with error { Option::None }, in which case we return empty data { lemme know if u think I am wrong here }.


</blockquote></details>

LGTM
___
*[`crates/evm/src/precompiles/identity.cairo` line 11 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm5RNEwBkjAhpy3OUTY:-Nm5RNEwBkjAhpy3OUTZ:b-khbifw) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/213b472e7eddd22dd2e2086d83e9ae9ecb669049/crates/evm/src/precompiles/identity.cairo#L11)):*
> ```Cairo
> #[generate_trait]
> impl IdentityPrecompileTraitImpl of IdentityPrecompileTrait {
>     fn address() -> EthAddress {
> ```

maybe address should be inlined though! it's quite small 
___
*[`crates/evm/src/tests/test_precompiles/test_ec_recover.cairo` line 19 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2GTz31x_pLwVS9ToH:-Nm5RS1O5Rn80KcGjQl3:b-kbgdmu) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/tests/test_precompiles/test_ec_recover.cairo#L19)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

Can a link to evm.codes be fine? Because I just took these from their playground, if this works, then I will add on the comment a link for the same!
</blockquote></details>

but the link won't be able to auto fill the calldata for precompile?

maybe explain how to reproduce the behaviour in detail

**Review by bajpai244 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675)* status: 1 of 6 files reviewed, 3 unresolved discussions (waiting on @ClementWalter and @Eikix)

___
*[`crates/evm/src/tests/test_precompiles/test_ec_recover.cairo` line 19 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2GTz31x_pLwVS9ToH:-Nm5X0Wq2g7B4j5KSdAX:b-d93bx0) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/tests/test_precompiles/test_ec_recover.cairo#L19)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

but the link won't be able to auto fill the calldata for precompile?

maybe explain how to reproduce the behaviour in detail
</blockquote></details>

So, the link opens up the whole test case as it is in the precompile, here is an example -> https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code='jFirsNplace_parameters%20in%20memoryZ456e9aea5e197a1f1af7a3e85a3212fa4049a3ba34c2289b4c860fc0b0c64ef3whash~Y~28wvX2YZ9242685bf161793cc25603c231bc2f568eb630ea16aa137d2664ac8038825608wrX4YZ4f8ae3bd7535248d0bd448298cc2e2071e56992d0774dc340c368ae950852adawsX6YqqjDo_call~32JSizeX80JOffsetX8VSize~VOffset~1waddressW4QFFFFFFFFwgasqSTATICCALLqqjPut_resulNalonKon_stackqPOPX80qMLOAD'~W1%20w%20jq%5Cnj//%20_%20thKZW32QY0qMSTOREX~0xWqPUSHV0wargsQ%200xNt%20Ke%20Jwret%01JKNQVWXYZ_jqw~_

Lemme know if I am missing something.

**Review by bajpai244 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675)* status: 1 of 6 files reviewed, 3 unresolved discussions (waiting on @ClementWalter and @Eikix)

___
*[`crates/evm/src/tests/test_precompiles/test_ec_recover.cairo` line 19 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm2GTz31x_pLwVS9ToH:-Nm5XFn_8hEaJl6abVN-:brb2uyf) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/f892356b5e8a532fdcd7b510e7e52e41ac7ce799/crates/evm/src/tests/test_precompiles/test_ec_recover.cairo#L19)):*
<details><summary><i>Previously, bajpai244 (Harsh Bajpai) wrote…</i></summary><blockquote>

So, the link opens up the whole test case as it is in the precompile, here is an example -> https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code='jFirsNplace_parameters%20in%20memoryZ456e9aea5e197a1f1af7a3e85a3212fa4049a3ba34c2289b4c860fc0b0c64ef3whash~Y~28wvX2YZ9242685bf161793cc25603c231bc2f568eb630ea16aa137d2664ac8038825608wrX4YZ4f8ae3bd7535248d0bd448298cc2e2071e56992d0774dc340c368ae950852adawsX6YqqjDo_call~32JSizeX80JOffsetX8VSize~VOffset~1waddressW4QFFFFFFFFwgasqSTATICCALLqqjPut_resulNalonKon_stackqPOPX80qMLOAD'~W1%20w%20jq%5Cnj//%20_%20thKZW32QY0qMSTOREX~0xWqPUSHV0wargsQ%200xNt%20Ke%20Jwret%01JKNQVWXYZ_jqw~_

Lemme know if I am missing something.
</blockquote></details>

But if you want, then I can also create a js script for this! Not against it in anyway, just want to make sure first that if u think that evm.codes link won't be good enough for this.

**Review by bajpai244 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675)* status: 1 of 6 files reviewed, 3 unresolved discussions (waiting on @ClementWalter and @Eikix)

___
*[`crates/evm/src/precompiles/identity.cairo` line 11 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-Nm5RNEwBkjAhpy3OUTY:-Nm5YOQ54mX7CsUpRDs7:b-7rl0jd) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/213b472e7eddd22dd2e2086d83e9ae9ecb669049/crates/evm/src/precompiles/identity.cairo#L11)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

maybe address should be inlined though! it's quite small
</blockquote></details>

done.

**Review by Eikix — DISMISSED** [review]
:lgtm:
:lgtm:

> Reviewed 3 of 5 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-:-Nm5h8uSDyWNyxy6ul4A:b-qcfkkb)* status: all files reviewed, 1 unresolved discussion (waiting on @bajpai244 and @ClementWalter)

**Review by bajpai244 — COMMENTED** [review]
@Eikix need the approval once again to merge :)

> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/675#-:-Nm5kuvi4qtjVMpsduaS:b-jb5gm2)* status: 5 of 7 files reviewed, 1 unresolved discussion (waiting on @ClementWalter and @Eikix)

File: crates/evm/src/tests/test_precompiles/test_identity.cairo [hunk 1]
```diff
@@ -8,6 +8,7 @@ use evm::stack::StackTrait;
 use evm::tests::test_utils::{VMBuilderTrait, native_token, other_starknet_address};
 use starknet::testing::set_contract_address;
 
+// source -> https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code='wFirsWplaceqparameters%20in%20memorybFFjdata~0vMSTOREvvwDoqcall~1QX3FQ_1YX1FY_4jaddressZ4%200xFFFFFFFFjgasvSTATICCALLvvwPutqresulWalonVonqstackvPOPb20vMLOAD'~Z1j//%20v%5Cnq%20thVj%20wb~0x_Offset~ZvPUSHYjargsXSizebWt%20Ve%20Qjret%01QVWXYZ_bjqvw~_
```
Comment (Eikix) [inline crates/evm/src/tests/test_precompiles/test_identity.cairo:11]:
```suggestion
// source: <https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code='wFirsWplaceqparameters%20in%20memorybFFjdata~0vMSTOREvvwDoqcall~1QX3FQ_1YX1FY_4jaddressZ4%200xFFFFFFFFjgasvSTATICCALLvvwPutqresulWalonVonqstackvPOPb20vMLOAD'~Z1j//%20v%5Cnq%20thVj%20wb~0x_Offset~ZvPUSHYjargsXSizebWt%20Ve%20Qjret%01QVWXYZ_bjqvw~_>
```

**Review by Eikix — APPROVED** [review]
lgtm

**Pull request merged**